### ISSUE TITLE
standardize Preconditioner and LinearOperator

### ISSUE THREAD

**Issue by ecoon** [amanzi/amanzi#454]
Preconditioner and LinearOperator are two different ways of getting an ApplyInverse() method.  There are currently two possible interpretations as to why these are different objects:

1, Preconditioners are approximate and LinearOperators are "exact to tolerance".
2, Preconditioners are assembled while LinearOperators are non-assembled (e.g. Krylov space methods)

But we have, for instance, LinearOperatorAmesos which is a direct solve LU, which is exact-to-tolerance, non-iterative, and assembled, while we also have PreconditionerDiagonal, which is approximate, non-iterative, and non-assembled.

So this would suggest 1 is correct, but the software design is really based around 2 -- Preconditioner has an Update() method which accepts an assembled Epetra_CrsMatrix.  LinearOperatorAmesos has to jump through a bunch of hoops to get an assembled matrix, because it doesn't have an Update() method.  Furthermore, it gets almost no reuse likely making it much less efficient, because it creates a new problem every single ApplyInverse (even if the matrix hasn't changed, e.g. in a linear problem).

I think we should move toward a more unified view of this, where we have one object -- I would suggest LinearOperator -- and we expand the interface to allow it to work with assembled objects.  Requirements on this would look something like:

1. Base class to allow using either e.g. Hypre or GMRES + ILU and storing them in a PK as the same thing (e.g. gets rid of dual preconditioner and linear operator objects in PKs.

2. You should be able to chain these -- pass a LinearOperatorILU as an argument to a LinearOperatorGMRES to get a preconditioned GMRES, for instance.  However, only the inner-most one may be assembled -- any outer LinearOperator in a chain cannot be assembled (e.g. it doesn't make sense to have LU wrapping a preconditioned thing).

3. Support separate forward and inverse operators (even though we currently don't use this).

4. Support the following specializations:

```
    LinearOperator<TreeOperator,TreeVector,TreeVectorSpace>
    LinearOperator<Operator,CompositeVector,CompositeVectorSpace>
    LinearOperator<Epetra_CrsMatrix, Epetra_Vector, Epetra_Map>
```

5. Support assembled and non-assembled methods, and doesn't assemble for non-assembled methods.

6. Explicitly and correctly support the "initialize()", "update() / compute()" and "applyInverse() / apply()" paradigm of Epetra / Tpetra respectively.  Don't recreate the object on every applyInverse() call to avoid throwing away reusable work.

**Comment by ecoon** [comment]
Note this relates to #453 as both will affect TreeOperators and their interface.

**Comment by lipnikov** [comment]
I would consider renaming LinearOperator to base Matrix and derived MatrixAssembled and MatrixPreconditioner. For solvers, I would suggest MethodIterativePCG, MethodDirectAmesos. Finally, PreconditionerBoomerAMG, and so on. Too many Operators around...

**Comment by ecoon** [comment]
I agree.  I think I'd prefer to go with Inverse over Matrix, as I'm reserving the name "Matrix" for use in PKs, where it can be used to mask the template parameters.  E.g.

`using Matrix = Inverse<Operator,CompositeVector,CompositeVectorSpace>;`

or similar.

**Comment by ecoon** [comment]
Question -- how do you feel about the following logic to parse the ParameterList passed to the InverseFactory?  (Order matters)

1. If the parameter "iterative method" appears -- require a "preconditioning method" parameter, construct and give the PC to the iterative method, and return the iterative method. **assumption: All iterative methods use PCs, even if just the identity**
2. If the parameter "direct method" appears -- construct the direct method, give the Operator to the direct method, and return the direct method. **assumption: No direct methods use PCs.**
3. If neither appears, but a "preconditioning method" parameter appears, then construct the PC and return it.
4. throw an error.

This implies that "iterative method" and "preconditioning method" appear in the same list (and NOT in the "diffusion" list, as is currently the case in ATS).

```
    <ParameterList name="solver">
      <Parameter name="iterative method" value="GMRES"/>
      <ParameterList name="GMRES parameters">
        ...
      </ParameterList>
      <Parameter name="preconditioning method" value="block ilu"/>
      <ParameterList name="block ilu parameters">
        ...
      </ParameterList>
    </ParameterList>
```

or

```
    <ParameterList name="solver">
      <Parameter name="direct method" value="amesos2: basker"/>
      <ParameterList name="amesos2: basker parameters">
        ...
      </ParameterList>
    </ParameterList>
```

or

```
    <ParameterList name="solver">
      <Parameter name="preconditioning method" value="boomer amg"/>
      <ParameterList name="boomer amg parameters">
        ...
      </ParameterList>
    </ParameterList>
```

**Comment by lipnikov** [comment]
Sounds reasonable to me.

**Comment by ecoon** [comment]
This was closed by #469
